pastis/temporal_analysis/harris_mode.ipynb

```diff
@@ -0,0 +1,3338 @@
+    "                                  norm):\n",
+    "    P = np.zeros(Q.shape)  # WFE modes covariance estimate\n",
+    "    r = Gsensor.shape[2]\n",
+    "    N = Gsensor.shape[0]\n",
+    "    N_img = Gcoro.shape[0]\n",
+    "    c = 1\n",
+    "    # Iterations of ALGORITHM 1\n",
+    "    contrast_hist = np.zeros(Niter)\n",
+    "    intensity_WFS_hist = np.zeros(Niter)\n",
+    "    cal_I_hist = np.zeros(Niter)\n",
+    "    eps_hist = np.zeros([Niter, r])\n",
+    "    averaged_hist = np.zeros(Niter)\n",
+    "    contrasts = []\n",
+    "    for pp in range(Niter):\n",
+    "        eps = np.random.multivariate_normal(np.zeros(r), P + Q * t_exp).reshape((1, 1, r))  # random modes\n",
+    "        G_eps = np.sum(Gsensor * eps, axis=2).reshape((N, 1, 2 * c)) + E0sensor  # electric field\n",
+    "        G_eps_squared = np.sum(G_eps * G_eps, axis=2, keepdims=True)\n",
+    "        G_eps_G = np.matmul(G_eps, Gsensor)\n",
+    "        G_eps_G_scaled = G_eps_G / np.sqrt(G_eps_squared + Dsensor / flux / t_exp)  # trick to save RAM\n",
+    "        cal_I = 4 * flux * t_exp * np.einsum(\"ijk,ijl->kl\", G_eps_G_scaled, G_eps_G_scaled)  # information matrix\n",
+    "        P = np.linalg.inv(np.linalg.inv(P + Q * t_exp / 2) + cal_I)\n",
+    "        #         P = np.linalg.inv(cal_I)\n",
+    "\n",
+    "        # Coronagraph\n",
+    "        G_eps_coron = np.sum(Gcoro * eps, axis=2).reshape((N_img, 1, 2 * c)) + E0coro\n",
+    "        G_eps_coron_squared = np.sum(G_eps_coron * G_eps_coron, axis=2, keepdims=True)\n",
+    "        intensity = G_eps_coron_squared * flux * t_exp + Dcoro\n",
+    "\n",
+    "        # Wavefront sensor\n",
+    "        intensity_WFS = G_eps_squared * flux * t_exp + Dsensor\n",
+    "\n",
+    "        # Archive\n",
+    "        test_DH0 = intensity[:, 0, 0] * luvoir.dh_mask\n",
+    "        test_DH = np.mean(test_DH0[np.where(test_DH0 != 0)])\n",
+    "        contrasts.append(test_DH / flux / t_exp / norm)\n",
+    "        intensity_WFS_hist[pp] = np.sum(intensity_WFS) / flux\n",
+    "        cal_I_hist[pp] = np.mean(cal_I) / flux\n",
+    "        eps_hist[pp] = eps\n",
+    "        averaged_hist[pp] = np.mean(contrasts)\n",
+    "        #         print(\"est. contrast\", np.mean(contrasts))\n",
+    "\n",
+    "        outputs = {'intensity_WFS_hist': intensity_WFS_hist,\n",
+    "                   'cal_I_hist': cal_I_hist,\n",
+    "                   'eps_hist': eps_hist,\n",
+    "                   'averaged_hist': averaged_hist,\n",
+    "                   'contrasts': contrasts}\n",
+    "    return outputs"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "entire-taiwan",
+   "metadata": {
+    "scrolled": true
+   },
+   "outputs": [],
+   "source": [
+    "def req_closedloop_calc_batch(Gcoro, Gsensor, E0coro, E0sensor, Dcoro, Dsensor, t_exp, flux, Q, Niter, dh_mask, norm):\n",
```

Comment:
This should be imported from 
https://github.com/spacetelescope/PASTIS/blob/dcd51b51fa3fa64c88e492b857623e49a97fc7e7/pastis/temporal_analysis/close_loop_analysis.py#L68
that's why we put it there. When you change this, make sure it still runs correctly please.